libs/mpi/include/hpx/mpi/mpi_future.hpp

```diff
@@ -0,0 +1,186 @@
+//  Copyright (c) 2019 John Biddiscombe
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef HPX_LIBS_MPI_FUTURE_HPP
+#define HPX_LIBS_MPI_FUTURE_HPP
+
+#include <mpi.h>
+//
+#include <cstdio>
+#include <iostream>
+#include <array>
+#include <utility>
+#include <tuple>
+#include <list>
+#include <mutex>
+//
+#include <hpx/local_lcos/promise.hpp>
+#include <hpx/util/yield_while.hpp>
+#include <hpx/memory/intrusive_ptr.hpp>
+#include <hpx/debugging/print.hpp>
+//
+
+namespace hpx { namespace mpi {
+
+    using  print_on = hpx::debug::enable_print<false>;
+    static print_on mpi_debug("MPI_FUTURE");
+
+    // mutex needed to protect mpi request list
+    static std::mutex list_mtx_;
+
+    // MPI request object list backed by mpi::future_data
+    struct future_data;
+    using  future_data_ptr = memory::intrusive_ptr<future_data>;
+    static std::list<future_data_ptr> active_requests_;
+
+    // -----------------------------------------------------------------
+    // An implementation of future_data for MPI
+    // -----------------------------------------------------------------
+    struct future_data : hpx::lcos::detail::future_data<void>
+    {
+        HPX_NON_COPYABLE(future_data);
+
+        using init_no_addref = hpx::lcos::detail::future_data<void>::init_no_addref;
+
+        // default empty constructor
+        future_data() = default;
+
+        // constructor that takes a request
+        future_data(init_no_addref no_addref, MPI_Request request)
+            : hpx::lcos::detail::future_data<void>(no_addref), request_(request)
+        {}
+
+        // constructor used for creation directly by invoke
+        future_data(init_no_addref no_addref)
+            : hpx::lcos::detail::future_data<void>(no_addref)
+        {}
+
+        // The native MPI request handle owned by this future data
+        MPI_Request request_;
+    };
+
+    // -----------------------------------------------------------------
+    // utility function to add a new request to the list to be tracked
+    void add_to_request_list(future_data_ptr data)
```

Comment:
Can this go into `namespace detail`?

Reply:
yes, this should be in detail, but I'd argue that the `poll` function is intended to be referenced by the end user so it should be in the top level namespace

Reply:
```
    sched->set_user_polling_function(&hpx::mpi::poll);
    sched->add_scheduler_mode(hpx::threads::policies::enable_user_polling);
```